idea/idea-fir/src/org/jetbrains/kotlin/idea/core/overrideImplement/KtGenerateMembersHandler.kt

```diff
@@ -0,0 +1,284 @@
+            for (existingDeclaration in existingDeclarations) {
+                val node: DoublyLinkedNode<MemberEntry> = DoublyLinkedNode(MemberEntry.ExistingEntry(existingDeclaration))
+                sentinelTailNode.prepend(node)
+                val callableSymbol = existingDeclaration.getSymbol() as? KtCallableSymbol ?: continue
+                for (overriddenSymbol in callableSymbol.getAllOverriddenSymbols()) {
+                    put(overriddenSymbol.psi ?: continue, node)
+                }
+            }
+        }
+
+        // Note on implementation: here we need the original ordering defined in the source code, so we stick to PSI rather than using
+        // `KtMemberScope` because FIR groups overloaded functions together.
+        outer@ for ((selectedSymbol, generatedPsi) in selectedMemberSymbols) {
+            val superSymbol = selectedSymbol.originalOverriddenSymbol
+            val superPsi = superSymbol?.psi
+            if (superPsi == null) {
+                // This normally should not happen, but we just try to play safe here.
+                sentinelTailNode.prepend(DoublyLinkedNode(MemberEntry.NewEntry(generatedPsi)))
+                continue
+            }
+            var currentPsi = superSymbol.psi?.prevSibling
+            while (currentPsi != null) {
+                val matchedNode = superPsiToMemberEntry[currentPsi]
+                if (matchedNode != null) {
+                    val newNode = DoublyLinkedNode<MemberEntry>(MemberEntry.NewEntry(generatedPsi))
+                    matchedNode.append(newNode)
+                    superPsiToMemberEntry[superPsi] = newNode
+                    continue@outer
+                }
+                currentPsi = currentPsi.prevSibling
+            }
+            currentPsi = superSymbol.psi?.nextSibling
+            while (currentPsi != null) {
+                val matchedNode = superPsiToMemberEntry[currentPsi]
+                if (matchedNode != null) {
+                    val newNode = DoublyLinkedNode<MemberEntry>(MemberEntry.NewEntry(generatedPsi))
+                    matchedNode.prepend(newNode)
+                    superPsiToMemberEntry[superPsi] = newNode
+                    continue@outer
+                }
+                currentPsi = currentPsi.nextSibling
+            }
+            val newNode = DoublyLinkedNode<MemberEntry>(MemberEntry.NewEntry(generatedPsi))
+            superPsiToMemberEntry[superPsi] = newNode
+            sentinelTailNode.prepend(newNode)
+        }
+        return sentinelHeadNode.toListSkippingNulls()
+    }
+
+    @OptIn(ExperimentalStdlibApi::class)
+    private fun insertMembersAccordingToPreferredOrder(
+        symbolsInPreferredOrder: List<MemberEntry>,
+        classLeftBrace: PsiElement?,
+        editor: Editor,
+        currentClass: KtClassOrObject
+    ): List<Iterable<KtDeclaration>> {
+        var firstAnchor: PsiElement? = null
+        if (symbolsInPreferredOrder.first() is MemberEntry.NewEntry) {
```

Comment:
The current logic guarantees it to be not empty because we have the empty check done in the beginning of `KtAnalysisSession.generateMembers`. Let me add a `require` check to make this intention explicit.